site/content/posts/2023-08-08-v0.25.0-impersonation-proxy-with-external-certs.md

```diff
@@ -0,0 +1,318 @@
+$ kubectl apply \
+    --filename concierge.webhookauthenticator.yaml \
+    --kubeconfig kind-with-contour.kubeconfig.yaml
+```
+
+Now deploy a Contour `HTTPProxy` ingress that fronts the `ClusterIP` service.
+Note in particular the `spec.tcpproxy` block, which is different than the typical `spec.rules` block. 
+`spec.tcpproxy` is required when using `spec.virtualhost.tls.passthrough: true`. 
+
+See https://projectcontour.io/docs/1.25/config/tls-termination/#tls-session-passthrough for more details.
+
+```shell
+$ cat << EOF > contour-ingress-impersonation-proxy.yaml
+---
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: impersonation-proxy
+  namespace: pinniped-concierge
+spec:
+  virtualhost:
+    fqdn: impersonation-proxy-mtls.local
+    tls:
+      passthrough: true
+  tcpproxy:
+    services:
+    - name: pinniped-concierge-impersonation-proxy-cluster-ip
+      port: 443
+EOF
+
+$ kubectl apply \
+    --filename contour-ingress-impersonation-proxy.yaml \
+    --kubeconfig kind-with-contour.kubeconfig.yaml
+```
+
+Now generate the Pinniped kubeconfig so that you can perform mTLS with the impersonation proxy.
+
+```shell
+# add 127.0.0.1 impersonation-proxy-mtls.local to your /etc/hosts!
+$ pinniped get kubeconfig \
+    --static-token "pinny:password123" \
+    --concierge-authenticator-type webhook \
+    --concierge-authenticator-name local-user-authenticator \
+    --concierge-mode ImpersonationProxy \
+    --kubeconfig kind-with-contour.kubeconfig.yaml \
+    > pinniped-kubeconfig.yaml
+```
+
+Now perform an action as user pinny!
+
+```shell
+$ kubectl get pods -A \
+    --kubeconfig pinniped-kubeconfig.yaml
+Error from server (Forbidden): pods is forbidden: User "pinny" cannot list resource "pods" in API group "" at the cluster scope: decision made by impersonation-proxy.concierge.pinniped.dev
+```
+
+This does result in an error because the cluster does not have any `RoleBindings` or `ClusterRoleBindings` that allow your user pinny or the group `group-for-mtls` to perform any actions on the cluster.
+Let’s make a `ClusterRoleBinding` that grants this group cluster admin privileges.
+
+```shell
+# Perform this as the cluster admin using the kind kubeconfig
+$ kubectl create clusterrolebinding mtls-admins \
+    --clusterrole=cluster-admin \
+    --group=group-for-mtls \
+    --kubeconfig kind-with-contour.kubeconfig.yaml
+# Now try again with the Pinniped kubeconfig
+$ kubectl get pods -A \
+    --kubeconfig pinniped-kubeconfig.yaml
+NAMESPACE                  NAME                                                      READY   STATUS      RESTARTS       AGE
+pinniped-concierge         pinniped-concierge-f4c78b674-bt6zl                        1/1     Running     0              3h36m
+```
+
+Congratulations, you have successfully performed mTLS authentication between your local client (kubectl, using the pinniped CLI) and the Pinniped concierge impersonation proxy inside the cluster.
+
```

Comment:
Shall we add a "going back to a generated cert" paragraph?

Reply:
No, I think the proposal has all of those details.